helix-core/src/main/java/org/apache/helix/controller/rebalancer/topology/Topology.java

```diff
@@ -212,123 +190,124 @@ private static Node cloneTree(Node root, Map<Node, Integer> newNodeWeight, Set<N
+      String instanceName, InstanceConfig instanceConfig, LinkedHashSet<String> clusterTopologyKeys,
+      Map<String, String> defaultDomainPathValuesCache) throws IllegalArgumentException {
+    LinkedHashMap<String, String> instanceTopologyMap = new LinkedHashMap<>();
+    if (isTopologyAwareEnabled) {
+      if (clusterTopologyKeys == null || clusterTopologyKeys.size() == 0) {
+        // Return a ordered map using default cluster topology definition, i,e. /root/zone/instance
+        String zone = instanceConfig.getZoneId();
+        if (zone == null) {
+          throw new IllegalArgumentException(String
+              .format("ZONE_ID for instance %s is not set, fail the topology-aware placement!",
+                  instanceName));
         }
-      }
-
-      String[] pathPairs = domain.trim().split(",");
-      Map<String, String> pathValueMap = new HashMap<>();
-      for (String pair : pathPairs) {
-        String[] values = pair.trim().split("=");
-        if (values.length != 2 || values[0].isEmpty() || values[1].isEmpty()) {
-          throw new HelixException(String.format(
-              "Domain-Value pair %s for instance %s is not valid, failed the topology-aware placement!",
-              pair, ins));
+        instanceTopologyMap.put(Types.ZONE.name(), zone);
+        instanceTopologyMap.put(Types.INSTANCE.name(), instanceName);
+      } else {
+        /*
+         * Return a ordered map representing the instance path. The topology order is defined in
+         * ClusterConfig.topology.
+         */
+        Map<String, String> domainAsMap;
+        try {
+          domainAsMap = instanceConfig.getDomainAsMap();
+        } catch (IllegalArgumentException e) {
+          throw new IllegalArgumentException(String
+              .format("Domain %s for instance %s is not valid, fail the topology-aware placement!",
+                  instanceConfig.getDomainAsString(), instanceName), e);
         }
-        String type = values[0];
-        String value = values[1];
 
-        if (!_types.contains(type)) {
-          logger.warn(String
-              .format("Path %s defined in domain of instance %s not recognized, ignored!", pair,
-                  ins));
-          continue;
+        if (domainAsMap == null || domainAsMap.isEmpty()) {
+          throw new IllegalArgumentException(String
+              .format("Domain for instance %s is not set, fail the topology-aware placement!",
+                  instanceName));
+        }
+        int numOfMatchedKeys = 0;
+        for (String key : clusterTopologyKeys) {
+          // if a key does not exist in the instance domain config, using the default domain value.
+          String value = domainAsMap.get(key);
+          if (value == null || value.length() == 0) {
+            value =
+                defaultDomainPathValuesCache.computeIfAbsent(key, k -> (DEFAULT_DOMAIN_PREFIX + key));
```

Comment:
I see you want a lazy initialization. It's good design.
But since you are modifying the input parameter implicitly, it might be clearer to just refer to _defaultDomainPathValues. Otherwise, there is an implicit assumption that the caller needs to pass the same map for all the calls.

Reply:
I changed it back to initialing the map in the constrictor. CPU profiler shows the overhead of having a lamda function here is too heavy.